### ISSUE TITLE
Dart and native wrapper client interfaces

### ISSUE THREAD

**Issue by bryanchriswhite** [LeastAuthority/dart_wormhole_william#7]
### Context

In order to interoperate between the dart and golang codebases, we must design the interface between them. As the dart ffi package is the officially supported path to native extensions, we will need to expose a C API to wormhole-william that dart can call into and receive callbacks from.

Considerations:
- Use idiomatic / conventional C error handling where appropriate
- Golang can have multiple return values
- Cgo might make it difficult to call C function pointers from dart ffi (e.g. `Pointer.fromFunction<T>(...)`).
- The go runtime will need to maintain a map of go objects which need to be persisted. The respective key is passed between dart and go to facilitate references to go objects from dart (e.g. exported C functions which are methods will receive this key as the first argument to lookup their receiver).

I suspect we will also need to move execution of the bound methods into a new dart isolate to parallelize the I/O heavy native extension code with the time-sensitive UI code (see: [\<TODO: issue link\>]()). 

#### Related links
- [cgo documentation](https://golang.org/cmd/cgo/)
- [Passing callbacks and pointers to cgo](https://eli.thegreenplace.net/2019/passing-callbacks-and-pointers-to-cgo/)
- [C interop using dart:ffi](https://dart.dev/guides/libraries/c-interop)

### Acceptance criteria:

- Interface specification fits within cgo and dart ffi constraints.
- API includes members to:
  - create a new client (e.g. `NewClient`)
  - garbage collect a client (e.g. `FreeClient`)
  - send a file (e.g. `Client_SendFile`)
  - receive a file (e.g. `Client_ReceiveFile`)

**Comment by bryanchriswhite** [comment]
I've put together a working example of asynchronous native callbacks to dart (including isolate_rpc integration) in this repo: [dart-ffi-callback-proof-of-concept](https://github.com/bryanchriswhite/dart-ffi-callback-proof-of-concept).

TLDR; 
- In order to callback to dart, we construct a `ReceivePort` on the dart-side and pass it's corresponding `SendPort` (`#sendPort.nativePort`) to the native function. The native function can then send a message using the linked dart native API dynamic library.
- To prevent blocking calls and general load on the main isolate, the example implements a class which contains native API calls to a separate isolate.
- To facilitate a request-response pattern on top of the message-passing API that isolates implement, the example integrates our [isolate_rpc](https://github.com/LeastAuthority/isolate_rpc) implementation (Dart port of [worker-rpc](https://github.com/DirtyHairy/worker-rpc)).
